### PR TITLE
Avoid SerDe for aggregation query with object pool

### PR DESCRIPTION
**Pull request by kangkaisen**
For https://github.com/apache/incubator-doris/pull/1610  "avoid serialize and deserialize from Scannode to Aggnode".

For aggregation query, we directly pass agg object from scan node to Agg node and avoid unnecessary serialize and deserialize.

### CODE REVIEW

File: be/src/olap/aggregate_func.h [hunk 1]
```diff
@@ -398,17 +401,21 @@ struct AggregateFuncTraits<OLAP_FIELD_AGGREGATION_REPLACE, OLAP_FIELD_TYPE_CHAR>
...
-        dst_slice->size = sizeof(HyperLogLog);
-        // use 'placement new' to allocate HyperLogLog on arena, so that we can control the memory usage.
-        char* mem = arena->Allocate(dst_slice->size);
-        dst_slice->data = (char*) new (mem) HyperLogLog((const uint8_t*)src_slice->data);
+        // we use int32_t max size represent this slice is a agg object
+        dst_slice->size = std::numeric_limits<int32_t>::max();
+        auto* hll = new HyperLogLog((const uint8_t*) src_slice->data);
+        dst_slice->data = reinterpret_cast<char*>(hll);
+
+        arena->track_memory(sizeof(HyperLogLog));
```
Comment (morningman) [inline be/src/olap/aggregate_func.h]: I think `sizeof(HyperLogLog)` does not reflect the real size of HLL, and even has great difference

**Review by morningman — CHANGES_REQUESTED** [review]

File: be/src/olap/aggregate_func.h (hunk 1 above)
Comment (kangkaisen) [reply be/src/olap/aggregate_func.h]: I agree with you.
But this should be another issue,  This PR don't change this point.
We need a better way to estimate the HLL and Bitmap memory usage.

File: be/src/olap/row_cursor.h [hunk 2]
```diff
@@ -161,6 +169,9 @@ class RowCursor {
     char* _variable_buf = nullptr;
     size_t _variable_len;
 
+    // for agg query, we don't need to finalize when scan agg object data
+    bool need_agg_finalize = true;
```
Comment (imay) [inline be/src/olap/row_cursor.h]: Is it a good option to put this flag here? This struct only a row, it should not care how to assemble itself

File: be/src/olap/reader.cpp [hunk 3]
```diff
@@ -389,11 +389,15 @@ OLAPStatus Reader::_agg_key_next_row(RowCursor* row_cursor, Arena* arena, bool*
         ++merged_count;
     } while (true);
     _merged_rows += merged_count;
-    agg_finalize_row(_value_cids, row_cursor, arena);
+    // For agg query, we don't need finalize agg object and directly pass agg object to agg node
+    if (row_cursor->is_need_agg_finalize()) {
```
Comment (imay) [inline be/src/olap/reader.cpp]: better to put flag in ReaderParams

File: be/src/olap/merger.cpp [hunk 4]
```diff
@@ -44,14 +44,14 @@ OLAPStatus Merger::merge_rowsets(TabletSharedPtr tablet,
     RETURN_NOT_OK_LOG(row_cursor.init(tablet->tablet_schema()),
                  "failed to init row cursor when merging rowsets of tablet " + tablet->full_name());
     row_cursor.allocate_memory_for_string_type(tablet->tablet_schema());
-
     // The following procedure would last for long time, half of one day, etc.
     int64_t output_rows = 0;
     while (true) {
         Arena arena;
+        ObjectPool objectPool;
```
Comment (imay) [inline be/src/olap/merger.cpp]: Wow, that's bad. we will allocate a new arena and object pool for each row we read. This is very bad for performance.

**Review by imay — CHANGES_REQUESTED** [review]

File: be/src/runtime/row_batch.cpp [hunk 5]
```diff
@@ -601,6 +607,7 @@ void RowBatch::swap(RowBatch* other) {
     std::swap(_has_in_flight_row, other->_has_in_flight_row);
     std::swap(_num_rows, other->_num_rows);
     std::swap(_capacity, other->_capacity);
+    _agg_object_pool.swap(other->_agg_object_pool);
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply be/src/runtime/row_batch.cpp]: OK

File: be/src/runtime/row_batch.cpp [hunk 6]
```diff
@@ -500,6 +505,7 @@ void RowBatch::close_tuple_streams() {
 void RowBatch::transfer_resource_ownership(RowBatch* dest) {
     dest->_auxiliary_mem_usage += _tuple_data_pool->total_allocated_bytes();
     dest->_tuple_data_pool->acquire_data(_tuple_data_pool.get(), false);
+    dest->_agg_object_pool.swap(_agg_object_pool);
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply be/src/runtime/row_batch.cpp]: has fixed

File: be/src/olap/schema_change.cpp [hunk 7]
```diff
@@ -578,7 +579,7 @@ bool RowBlockMerger::merge(
 
     row_cursor.allocate_memory_for_string_type(_tablet->tablet_schema());
     while (_heap.size() > 0) {
-        init_row_with_others(&row_cursor, *(_heap.top().row_cursor), arena.get());
+        init_row_with_others(&row_cursor, *(_heap.top().row_cursor), arena.get(), agg_object_pool.get());
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply be/src/olap/schema_change.cpp]: OK

File: be/src/olap/task/engine_checksum_task.cpp [hunk 8]
```diff
@@ -115,7 +116,7 @@ OLAPStatus EngineChecksumTask::_compute_checksum() {
     bool eof = false;
     uint32_t row_checksum = 0;
     while (true) {
-        OLAPStatus res = reader.next_row_with_aggregation(&row, arena.get(), &eof);
+        OLAPStatus res = reader.next_row_with_aggregation(&row, arena.get(), agg_object_pool.get(), &eof);
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply be/src/olap/task/engine_checksum_task.cpp]: OK

File: be/src/olap/reader.cpp (hunk 3 above)
Comment (kangkaisen) [reply be/src/olap/reader.cpp]: OK

File: be/src/olap/merger.cpp (hunk 4 above)
Comment (kangkaisen) [reply be/src/olap/merger.cpp]: I will fix this issue in another PR.

**Comment by imay** [comment]
@kangkaisen 

I suddenly remembered that you need to modify HLL related functions in be/src/exprs/aggregate_functions.cpp. Otherwise, it will cause BE Crash during the upgrade process.

**Comment by kangkaisen** [comment]
> @kangkaisen
> 
> I suddenly remembered that you need to modify HLL related functions in be/src/exprs/aggregate_functions.cpp. Otherwise, it will cause BE Crash during the upgrade process.

Thanks for your reminder. update.

File: be/src/udf/udf.cpp [hunk 9]
```diff
@@ -465,9 +465,19 @@ void HllVal::init(FunctionContext* ctx) {
...
 
-void HllVal::agg_parse_and_cal(const HllVal &other) {
+void HllVal::agg_parse_and_cal(FunctionContext* ctx, const HllVal& other) {
     doris::HllSetResolver resolver;
-    resolver.init((char*)other.ptr, other.len);
+
+    // zero size means the src input is a HyperLogLog object
+    if (other.len == 0) {
+        auto* hll = reinterpret_cast<doris::HyperLogLog*>(other.ptr);
+        uint8_t* ptr = ctx->allocate(doris::HLL_COLUMN_DEFAULT_LEN);
```
Comment (imay) [inline be/src/udf/udf.cpp]: Because there is a private member variable named ptr and len, better to rename this ptr to other_ptr.

Comment (kangkaisen) [reply be/src/udf/udf.cpp]: done

**Review by imay — APPROVED** [review]
LGTM

File: be/src/olap/aggregate_func.h (hunk 1 above)
Comment (morningman) [reply be/src/olap/aggregate_func.h]: ok

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**